### PR TITLE
added SCRProcessorCreator

### PR DESCRIPTION
**Pull request by raduchis**
## Reasoning behind the pull request
- SCRProcessor is also created differently on normal and sovereign run type
  
## Proposed changes
- Implemented a factory for SCProcessor creation

## Testing procedure
- #allIn

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: factory/runType/runTypeComponentsHandler.go [hunk 1]
```diff
@@ -265,6 +269,18 @@ func (mrc *managedRunTypeComponents) ValidatorStatisticsProcessorCreator() peer.
 	return mrc.runTypeComponents.validatorStatisticsProcessorCreator
 }
 
+// SCProcessorCreator returns the smart contract processor creator
+func (mrc *managedRunTypeComponents) SCProcessorCreator() scrCommon.SCProcessorCreator {
+	mrc.mutStateComponents.RLock()
+	defer mrc.mutStateComponents.RUnlock()
+
+	if mrc.runTypeComponents == nil {
```
Comment (sasurobert) [inline factory/runType/runTypeComponentsHandler.go:277]: checl.ifnil ?

**Review by sasurobert — APPROVED** [review]

File: factory/runType/runTypeComponentsHandler.go (hunk 1 above)
Comment (sasurobert) [reply factory/runType/runTypeComponentsHandler.go:277]: on every if statement which looks like that.

File: process/smartContract/processorV2/sovereignSCProcess.go [hunk 2]
```diff
@@ -12,19 +12,19 @@ import (
 )
 
 type sovereignSCProcessor struct {
-	*scProcessor
+	process.SmartContractProcessorFacade
 }
 
 // TODO: use scrProcessorV2 when feat/vm1.5 is merged into feat/chain-sdk-go
```
Comment (mariusmihaic) [inline process/smartContract/processorV2/sovereignSCProcess.go:18]: We can delete this todo

File: process/interface.go [hunk 3]
```diff
@@ -65,6 +65,11 @@ type SmartContractResultProcessor interface {
 type SmartContractProcessorFacade interface {
 	SmartContractProcessor
 	SmartContractResultProcessor
+
+	ArgsParser() ArgumentsParser
+	TxTypeHandler() TxTypeHandler
```
Comment (mariusmihaic) [inline process/interface.go:70]: Not sure I understand this correctly, but why do we need these? 
These are the components that each scr processor receives on constructor, uses internally and are the same for each of them.

Comment (mariusmihaic) [reply process/interface.go:70]: This also applies to `CheckSCRBeforeProcessing`. This should be a separate small component that each scr processor receives and uses internally. No need to duplicate the code inside each scr processor.

**Review by mariusmihaic — COMMENTED** [review]
Before merging this this into feat, then the feat intro `feat/chain-go-sdk`, please test that the system test functionality is working with the mockNotifier

File: factory/runType/runTypeComponentsHandler.go (hunk 1 above)
Comment (raduchis) [reply factory/runType/runTypeComponentsHandler.go:277]: fixed

File: factory/runType/runTypeComponents.go [hunk 4]
```diff
@@ -104,6 +107,11 @@ func (rcf *runTypeComponentsFactory) Create() (*runTypeComponents, error) {
 		return nil, fmt.Errorf("runTypeComponentsFactory - NewShardAdditionalStorageServiceFactory failed: %w", err)
 	}
 
+	sCProcessorCreator, err := processProxy.NewSCProcessProxyFactory()
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/runType/runTypeComponents.go:110]: fixed

File: factory/processing/blockProcessorCreator.go [hunk 5]
```diff
@@ -257,7 +256,7 @@ func (pcf *processComponentsFactory) newShardBlockProcessor(
 		VMOutputCacher:      txcache.NewDisabledCache(),
 		WasmVMChangeLocker:  wasmVMChangeLocker,
 	}
-	scProcessorProxy, err := processorV2.CreateSCRProcessor(pcf.chainRunType, argsNewScProcessor)
+	scProcessorProxy, err := pcf.runTypeComponents.SCProcessorCreator().CreateSCProcessor(argsNewScProcessor, pcf.epochNotifier)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/processing/blockProcessorCreator.go:259]: fixed, added in the args

File: process/interface.go (hunk 3 above)
Comment (raduchis) [reply process/interface.go:70]: needed an interface to pass to the sovereignSCProcessorCreator, that could take either a scprocessor, a scprocessorV2, or a scProcessorProxy. 

CheckSCRBeforeProcessing was internal and made it public to be able to use it from the new interface.

Comment (raduchis) [reply process/interface.go]: refactored to remove the ArgsParser() and TxTypeHandler()

Comment (raduchis) [reply process/interface.go]: removed also CheckSCRBeforeProcessing

File: process/smartContract/scrCommon/scProcessHelper.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+type scProcessHelper struct {
+	accounts        state.AccountsAdapter
+	coordinator     sharding.Coordinator
+	marshalizer     marshal.Marshalizer
+	hasher          hashing.Hasher
+	pubKeyConverter core.PubkeyConverter
+}
+
+// SCProcessHelperArgs represents the arguments needed to create a new accounts helper
```
Comment (mariusmihaic) [inline process/smartContract/scrCommon/scProcessHelper.go:25]:
```suggestion
// SCProcessHelperArgs represents the arguments needed to create a new scr processor helper
```

File: process/smartContract/scrCommon/scProcessHelper_test.go [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+func TestSCProcessHelper_CheckSCRBeforeProcessing(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil SCR should not err", func(t *testing.T) {
+		args := getSCProcessHelperArgs()
+		sch, _ := scrCommon.NewSCProcessHelper(args)
+
+		account, err := sch.CheckSCRBeforeProcessing(nil)
```
Comment (mariusmihaic) [inline process/smartContract/scrCommon/scProcessHelper_test.go:152]: Here and everywhere below, the result of this func is not an `account`

File: process/smartContract/scrCommon/scProcessHelper_test.go [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+		}
+		sch, _ := scrCommon.NewSCProcessHelper(args)
+
+		scr := &smartContractResult.SmartContractResult{
+			Nonce:   0,
+			RcvAddr: []byte{1},
+			SndAddr: []byte{1},
+		}
+		account, err := sch.CheckSCRBeforeProcessing(scr)
+		require.NotNil(t, account)
```
Comment (mariusmihaic) [inline process/smartContract/scrCommon/scProcessHelper_test.go:245]: Perhaps check some data from this result to be valid?

File: process/errors.go [hunk 9]
```diff
@@ -1250,3 +1250,6 @@ var ErrNilPreProcessorCreator = errors.New("nil pre processor creator")
 
 // ErrNilSCProcessorCreator signals that a nil sc processor creator was provided
 var ErrNilSCProcessorCreator = errors.New("nil sc processor creator")
+
+// ErrNilSCProcessHelper signals that a nil sc processor helper was provided
+var ErrNilSCProcessHelper = errors.New("nil sc processor helper")
```
[the remark this answers is not in the document]
Comment (raduchis) [reply process/errors.go:1255]: Processor

File: process/interface.go [hunk 10]
```diff
@@ -1364,3 +1359,10 @@ type ScrProcessingDataHandler interface {
 	GetSender() state.UserAccountHandler
 	GetDestination() state.UserAccountHandler
 }
+
+// SCProcessHelperHandler is an interface for smart contract process helper
+type SCProcessHelperHandler interface {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply process/interface.go:1364]: Processor

File: process/smartContract/process.go [hunk 11]
```diff
@@ -2714,23 +2728,7 @@ func (sc *scProcessor) deleteAccounts(deletedAccounts [][]byte) error {
 }
 
 func (sc *scProcessor) getAccountFromAddress(address []byte) (state.UserAccountHandler, error) {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply process/smartContract/process.go:2730]: yes, removed

File: process/smartContract/scrCommon/scProcessHelper.go (hunk 6 above)
Comment (raduchis) [reply process/smartContract/scrCommon/scProcessHelper.go:25]: renamed

File: process/smartContract/processorV2/processV2.go [hunk 12]
```diff
@@ -203,9 +217,9 @@ func NewSmartContractProcessorV2(args scrCommon.ArgsNewSmartContractProcessor) (
 		storePerByte:        baseOperationCost["StorePerByte"],
 		persistPerByte:      baseOperationCost["PersistPerByte"],
 		executableCheckers:  scrCommon.CreateExecutableCheckersMap(args.BuiltInFunctions),
+		scProcessHelper:     scProcessHelper,
```
[the remark this answers is not in the document]
Comment (raduchis) [reply process/smartContract/processorV2/processV2.go:220]: didn't find any `getAccount` in this file

File: process/smartContract/scrCommon/scProcessHelper_test.go (hunk 7 above)
Comment (raduchis) [reply process/smartContract/scrCommon/scProcessHelper_test.go:152]: fixed, renamed to scrData

File: process/smartContract/scrCommon/scProcessHelper_test.go (hunk 8 above)
Comment (raduchis) [reply process/smartContract/scrCommon/scProcessHelper_test.go:245]: added test for actual values

**Review by mariusmihaic — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**